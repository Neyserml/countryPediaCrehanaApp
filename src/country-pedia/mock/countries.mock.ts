import { Country } from "../domain/country";
import { GeographicSubRegion } from "../domain/country-pedia-response";

export const COUNTRIES_STUB: Country[] = [
  {
    name: "Afghanistan",
    area: 652230,
    capital: "Kabul",
    population: 27657145,
    alpha2Code: "AF",
    nativeName: "افغانستان",
    flag: {
      emoji: "🇦🇫",
      emojiUnicode: "U+1F1E6 U+1F1EB",
      svgFile: "https://restcountries.eu/data/afg.svg",
    },
    location: {
      latitude: 33,
      longitude: 65,
    },
    currencies: [
      {
        name: "Afghan afghani",
        symbol: "؋",
      },
    ],
    officialLanguages: [
      {
        name: "Turkmen",
        nativeName: "Türkmen",
      },
      {
        name: "Uzbek",
        nativeName: "Oʻzbek",
      },
      {
        name: "Pashto",
        nativeName: "پښتو",
      },
    ],
    callingCodes: [
      {
        name: "93",
      },
    ],
    subregion: {
      name: "Southern Asia",
      region: {
        name: "Asia",
      },
    } as GeographicSubRegion,
  },
  {
    name: "Albania",
    area: 28748,
    capital: "Tirana",
    population: 2886026,
    alpha2Code: "AL",
    nativeName: "Shqipëria",
    flag: {
      emoji: "🇦🇱",
      emojiUnicode: "U+1F1E6 U+1F1F1",
      svgFile: "https://restcountries.eu/data/alb.svg",
    },
    location: {
      latitude: 41,
      longitude: 20,
    },
    currencies: [
      {
        name: "Albanian lek",
        symbol: "L",
      },
    ],
    officialLanguages: [
      {
        name: "Albanian",
        nativeName: "Shqip",
      },
    ],
    callingCodes: [
      {
        name: "355",
      },
    ],
    subregion: {
      name: "Southern Europe",
      region: {
        name: "Europe",
      },
    } as GeographicSubRegion,
  },
  {
    name: "Angola",
    area: 1246700,
    capital: "Luanda",
    population: 25868000,
    alpha2Code: "AO",
    nativeName: "Angola",
    flag: {
      emoji: "🇦🇴",
      emojiUnicode: "U+1F1E6 U+1F1F4",
      svgFile: "https://restcountries.eu/data/ago.svg",
    },
    location: {
      latitude: -12.5,
      longitude: 18.5,
    },
    currencies: [
      {
        name: "Angolan kwanza",
        symbol: "Kz",
      },
    ],
    officialLanguages: [
      {
        name: "Portuguese",
        nativeName: "Português",
      },
    ],
    callingCodes: [
      {
        name: "244",
      },
    ],
    subregion: {
      name: "Middle Africa",
      region: {
        name: "Africa",
      },
    } as GeographicSubRegion,
  },
];
