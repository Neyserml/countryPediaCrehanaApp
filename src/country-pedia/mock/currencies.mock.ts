import { Currency } from "../domain/country-pedia-response";

export const CURRENCIES_STUB: Currency[] = [
  {
    name: "Afghan afghani",
    symbol: "؋",
    code: "AFN",
  },
  {
    name: "Euro",
    symbol: "€",
    code: "EUR",
  },
  {
    name: "Albanian lek",
    symbol: "L",
    code: "ALL",
  },
  {
    name: "Algerian dinar",
    symbol: "د.ج",
    code: "DZD",
  },
  {
    name: "United State Dollar",
    symbol: "$",
    code: "USD",
  },
  {
    name: "Angolan kwanza",
    symbol: "Kz",
    code: "AOA",
  },
  {
    name: "East Caribbean dollar",
    symbol: "$",
    code: "XCD",
  },
  {
    name: "Australian dollar",
    symbol: "$",
    code: "AUD",
  },
  {
    name: "British pound",
    symbol: "£",
    code: "GBP",
  },
];
