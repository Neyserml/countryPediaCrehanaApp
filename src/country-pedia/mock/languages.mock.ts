import { Language } from "../domain/country-pedia-response";

export const LANGUAGES_STUB: Language[] = [
  {
    name: "Pashto",
    nativeName: "پښتو",
  },
  {
    name: "Uzbek",
    nativeName: "Oʻzbek",
  },
  {
    name: "Turkmen",
    nativeName: "Türkmen",
  },
  {
    name: "Swedish",
    nativeName: "svenska",
  },
  {
    name: "Albanian",
    nativeName: "Shqip",
  },
  {
    name: "Arabic",
    nativeName: "العربية",
  },
  {
    name: "English",
    nativeName: "English",
  },
  {
    name: "Samoan",
    nativeName: "gagana fa'a Samoa",
  },
  {
    name: "Catalan",
    nativeName: "català",
  },
  {
    name: "Portuguese",
    nativeName: "Português",
  },
];
