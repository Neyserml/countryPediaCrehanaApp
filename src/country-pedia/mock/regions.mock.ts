import { Region } from "../domain/country-pedia-response";

export const REGIONS_STUB: Region[] = [
  {
    name: "Asia",
    subregions: [
      {
        name: "Central Asia",
      },
      {
        name: "Eastern Asia",
      },
      {
        name: "South-Eastern Asia",
      },
      {
        name: "Western Asia",
      },
      {
        name: "Southern Asia",
      },
    ],
  },
  {
    name: "Europe",
    subregions: [
      {
        name: "Eastern Europe",
      },
      {
        name: "Western Europe",
      },
      {
        name: "Southern Europe",
      },
      {
        name: "Northern Europe",
      },
    ],
  },
  {
    name: "Africa",
    subregions: [
      {
        name: "Eastern Africa",
      },
      {
        name: "Southern Africa",
      },
      {
        name: "Western Africa",
      },
      {
        name: "Middle Africa",
      },
      {
        name: "Northern Africa",
      },
    ],
  },
  {
    name: "Oceania",
    subregions: [
      {
        name: "Micronesia",
      },
      {
        name: "Melanesia",
      },
      {
        name: "Australia and New Zealand",
      },
      {
        name: "Polynesia",
      },
    ],
  },
  {
    name: "Americas",
    subregions: [
      {
        name: "Northern America",
      },
      {
        name: "Central America",
      },
      {
        name: "South America",
      },
      {
        name: "Caribbean",
      },
    ],
  },
  {
    name: "Polar",
    subregions: [
      {
        name: "",
      },
    ],
  },
];
