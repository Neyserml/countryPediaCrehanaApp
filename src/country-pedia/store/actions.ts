interface DispatchObject<T, P = any> {
  type: T;
  payload?: P;
}

export enum CountryPediaAction {
  ClearState = "CLEAR_STATE",
  SetLanguages = "SET_LANGUAGES",
  SetCurrencies = "SET_CURRENCIES",
  SetRegions = "SET_REGIONS",
  SetCountries = "SET_COUNTRIES",
  SetCountryId = "SET_COUNTRY_ID",
}

export type CountryPediaDispatchObject<T> = DispatchObject<
  CountryPediaAction,
  T
>;
