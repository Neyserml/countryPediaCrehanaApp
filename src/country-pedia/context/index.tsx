import { castDraft } from "immer";
import React, { createContext, PropsWithChildren, useReducer } from "react";
import { DispatchState } from "../../shared/hooks/useDispatchState";
import { CountryPediaDispatchObject } from "../store/actions";
import {
  countryPediaContextReducer,
  CountryPediaState,
  initialState,
} from "../store/reducer";

export type CountryPediaDispatch<T = any> = React.Dispatch<
  CountryPediaDispatchObject<T>
>;

export type CountryPediaContextType = DispatchState<
  CountryPediaState,
  CountryPediaDispatch
>;

export const CountryPediaContext = createContext({} as CountryPediaContextType);

type CountryPediaProviderProps = PropsWithChildren<{
  extraState?: Partial<CountryPediaState>;
}>;

export const CountryPediaProvider: React.FC<CountryPediaProviderProps> = ({
  children,
  extraState = {},
}: CountryPediaProviderProps) => {
  const [state, dispatch] = useReducer(countryPediaContextReducer, {
    ...initialState,
    ...extraState,
  });

  const value = { dispatch, state: castDraft(state) };
  return (
    <CountryPediaContext.Provider value={value}>
      {children}
    </CountryPediaContext.Provider>
  );
};
