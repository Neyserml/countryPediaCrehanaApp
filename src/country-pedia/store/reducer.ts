import produce, { Draft } from "immer";
import { Country } from "../domain/country";
import { Currency, Language, Region } from "../domain/country-pedia-response";
import { CountryPediaAction, CountryPediaDispatchObject } from "./actions";

export interface CountryPediaState {
  title: string;
  languages: Language[] | null;
  currencies: Currency[] | null;
  regions: Region[] | null;
  countries: Country[] | null;
  countryId: string;
}

export const initialState: CountryPediaState = {
  title: "CountryPedia",
  languages: null,
  currencies: null,
  regions: null,
  countries: null,
  countryId: "",
};

export const countryPediaContextReducer = produce(
  (
    draft: Draft<CountryPediaState>,
    { type, payload }: CountryPediaDispatchObject<any>
  ) => {
    switch (type) {
      case CountryPediaAction.ClearState:
        return initialState;
      case CountryPediaAction.SetLanguages:
        draft.languages = payload;
        break;
      case CountryPediaAction.SetCurrencies:
        draft.currencies = payload;
        break;
      case CountryPediaAction.SetRegions:
        draft.regions = payload;
        break;
      case CountryPediaAction.SetCountries:
        draft.countries = payload;
        break;
      case CountryPediaAction.SetCountryId:
        draft.countryId = payload;
        break;
      default:
        throw new Error(`Unknown countryPedia action type: ${type}`);
    }
  }
);
