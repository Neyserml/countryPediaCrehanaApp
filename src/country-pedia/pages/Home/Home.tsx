import React, { ChangeEvent, useEffect, useState } from "react";
import { useForm } from "react-hook-form";
import { useHistory } from "react-router-dom";
import styled from "styled-components";
import CardTemplate from "../../../shared/components/CardTemplate/CardTemplete";
import Header from "../../../shared/components/Header/Header";
import useDispatchState from "../../../shared/hooks/useDispatchState";
import { CountryPediaContext } from "../../context";
import { Country } from "../../domain/country";
import {
  Currency,
  Language,
  Region,
} from "../../domain/country-pedia-response";
import { ApolloClientCountryPediaRequest } from "../../proxy/apollo-client-country-pedia.request";
import { CountryPediaAction } from "../../store/actions";
import { COUNTRY_PEDIA_MODULE_ROUTES as Pages } from "../../routes";

const StayledHome = styled.div`
  h3 {
    text-align: center;
  }

  label {
    margin-top: 10px;
    color: red;
  }
`;

const Home: React.FC = () => {
  const [state, dispatch] = useDispatchState(CountryPediaContext);
  const [search, setSearch] = useState({ value: "", touched: false });
  const [countriesFiltered, setCountriesFiltered] = useState<Country[] | null>(
    []
  );
  const { register, handleSubmit, reset } = useForm({
    mode: "all",
  });

  const { languages } = ApolloClientCountryPediaRequest.useLanguagesQuery();
  const { currencies } = ApolloClientCountryPediaRequest.useCurrenciesQuery();
  const { regions } = ApolloClientCountryPediaRequest.useRegionsQuery();
  const {
    countries,
    loading,
  } = ApolloClientCountryPediaRequest.useCountriesQuery();
  const history = useHistory();

  useEffect(() => {
    setCountriesFiltered(countries);
    dispatch({ type: CountryPediaAction.SetLanguages, payload: languages });
    dispatch({
      type: CountryPediaAction.SetCurrencies,
      payload: currencies?.filter((currency) => currency.name !== "null"),
    });
    dispatch({ type: CountryPediaAction.SetRegions, payload: regions });
    dispatch({ type: CountryPediaAction.SetCountries, payload: countries });
  }, [languages, currencies, regions, countries]);

  useEffect(() => {
    const searchUpperCase = search.value.toUpperCase();
    const resultFilterSearch = state.countries?.filter(
      ({ name, alpha2Code }) =>
        searchUpperCase === name.toUpperCase() || searchUpperCase === alpha2Code
    );
    if (resultFilterSearch && search.touched && search.value.length) {
      setCountriesFiltered(resultFilterSearch);
    } else {
      setCountriesFiltered(state.countries);
    }
  }, [state.countries, search.value]);

  const onFilter = ({
    language,
    currency,
    region,
  }: {
    language: string;
    currency: string;
    region: string;
  }) => {
    if (language) {
      const resultLanguage = filterByLanguage(language) as Country[];
      setCountriesFiltered(resultLanguage);
    }
    if (currency) {
      const resultCurrency = filterByCurrency(currency) as Country[];
      setCountriesFiltered(resultCurrency);
    }

    if (region) {
      const resultRegions = filterByRegion(region) as Country[];
      setCountriesFiltered(resultRegions);
    }
  };

  const filterByLanguage = (language: string): Country[] | undefined => {
    const result = countriesFiltered?.filter(({ officialLanguages }: Country) =>
      officialLanguages?.some(({ name }) => name === language)
    );

    return result;
  };

  const filterByCurrency = (currency: string) => {
    const result = countriesFiltered?.filter(({ currencies }) =>
      currencies?.some(({ name }) => name === currency)
    );

    return result;
  };

  const filterByRegion = (region: string) => {
    return countriesFiltered?.filter(
      ({ subregion }) => subregion?.region.name === region
    );
  };

  const onClear = (): void => {
    reset({ language: "", currency: "", region: "" });
    setCountriesFiltered(state.countries);
  };

  const onDetail = (name: string): void => {
    dispatch({ type: CountryPediaAction.SetCountryId, payload: name });
    history.push(Pages.Detail);
  };

  return (
    <StayledHome>
      <Header />
      <h3>CountryPedia</h3>
      <div className="container">
        <input
          value={search.value}
          onBlur={() => {
            setSearch({ value: "", touched: false });
          }}
          onChange={(e: ChangeEvent<HTMLInputElement>) => {
            setSearch({ value: e.target.value, touched: true });
          }}
          placeholder="Ingrese nombre código de pais a buscar"
          type="text"
        />
        <h6>Filtrar por:</h6>
        <div className="row">
          <div className="col s12 m4">
            <select
              name="language"
              onChange={handleSubmit(onFilter)}
              ref={register}
              className="browser-default"
              defaultValue=""
            >
              <option value="" disabled>
                Idioma
              </option>
              {state.languages?.map(({ name, nativeName }: Language) => (
                <option key={name + nativeName} value={name}>
                  {name}
                </option>
              ))}
            </select>
          </div>
          <div className="col s12 m4">
            <select
              name="currency"
              onChange={handleSubmit(onFilter)}
              ref={register}
              className="browser-default"
              defaultValue=""
            >
              <option value="" disabled>
                Moneda
              </option>
              {state.currencies?.map(({ name, code }: Currency) => (
                <option key={code + name} value={name}>
                  {name}
                </option>
              ))}
            </select>
          </div>
          <div className="col s12 m4">
            <select
              name="region"
              onChange={handleSubmit(onFilter)}
              ref={register}
              className="browser-default"
              defaultValue=""
            >
              <option value="" disabled>
                Región
              </option>
              {state.regions?.map(({ name }: Region) => (
                <option key={name} value={name}>
                  {name}
                </option>
              ))}
            </select>
          </div>
        </div>
        <button
          className="btn waves-effect waves-light red"
          type="submit"
          onClick={handleSubmit(onClear)}
          name="action"
        >
          Limpiar
        </button>

        <div className="row">
          {loading && (
            <div className="preloader-wrapper big active">
              <div className="spinner-layer spinner-blue-only">
                <div className="circle-clipper left">
                  <div className="circle"></div>
                </div>
              </div>
            </div>
          )}
          {countriesFiltered?.length === 0 && (
            <label>No se encontraron resultados.</label>
          )}

          {countriesFiltered &&
            countriesFiltered.map((country: Country) => (
              <div key={country.name} className="col s12 m4">
                <CardTemplate
                  onClickDetail={onDetail}
                  title={country.name}
                  img={country.flag.svgFile}
                >
                  <b>Region:</b> {country.subregion?.region?.name}
                  <br />
                  <b>Moneda:</b>{" "}
                  {country.currencies.map(({ name }, index: number) => (
                    <React.Fragment key={name + index}>
                      {name}
                      {index !== country.currencies.length - 1 ? "," : null}
                    </React.Fragment>
                  ))}
                  <br />
                  <b>Idioma:</b>
                  {country.officialLanguages.map(({ name }, index: number) => (
                    <React.Fragment key={name + index}>
                      {name}
                      {index !== country.officialLanguages.length - 1
                        ? ","
                        : null}
                    </React.Fragment>
                  ))}
                  <br />
                </CardTemplate>
              </div>
            ))}
        </div>
      </div>
    </StayledHome>
  );
};

export default Home;
