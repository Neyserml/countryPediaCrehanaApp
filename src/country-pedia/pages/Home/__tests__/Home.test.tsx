import { act, fireEvent, render, screen } from "@testing-library/react";
import React from "react";
import ApolloConnectedProvider from "../../../../config";
import { CountryPediaProvider } from "../../../context";
import { COUNTRIES_STUB } from "../../../mock/countries.mock";
import { CURRENCIES_STUB } from "../../../mock/currencies.mock";
import { LANGUAGES_STUB } from "../../../mock/languages.mock";
import { REGIONS_STUB } from "../../../mock/regions.mock";
import { CountryPediaState, initialState } from "../../../store/reducer";
import Home from "../Home";

const state: CountryPediaState = {
  ...initialState,
  countries: COUNTRIES_STUB,
  languages: LANGUAGES_STUB,
  currencies: CURRENCIES_STUB,
  regions: REGIONS_STUB,
};

const renderComponent = (extraState = state) => {
  return render(
    <ApolloConnectedProvider>
      <CountryPediaProvider extraState={extraState}>
        <Home />
      </CountryPediaProvider>
    </ApolloConnectedProvider>
  );
};

describe("Home Component", () => {
  beforeEach(() => {
    jest.useFakeTimers();
    jest.clearAllMocks();
  });

  it("should match snapshot in home page", async () => {
    await act(async () => {
      const component = renderComponent();
      expect(component.container).toMatchSnapshot();
    });
  });

  it("should filter according to name entered", async () => {
    await act(async () => {
      renderComponent();
    });
    const name = screen.queryByPlaceholderText(
      "Ingrese nombre c√≥digo de pais a buscar"
    ) as HTMLInputElement;

    fireEvent.change(name, { target: { value: "Afghanistan" } });

    const message = screen.queryByText("No se encontraron resultados.");
    expect(message).toBeNull();
  });
});
