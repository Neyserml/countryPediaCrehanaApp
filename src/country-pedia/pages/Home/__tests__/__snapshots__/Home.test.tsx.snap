// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Component should match snapshot in home page 1`] = `
<div>
  <div
    class="sc-hKgILt bpcmEC"
  >
    <nav
      class="sc-dlfnbm etGDIC"
    >
      <nav
        class="nav-wrapper"
      >
        <a
          class="brand-logo"
          href="#"
        >
          Crehana
        </a>
      </nav>
    </nav>
    <h3>
      CountryPedia
    </h3>
    <div
      class="container"
    >
      <input
        placeholder="Ingrese nombre código de pais a buscar"
        type="text"
        value=""
      />
      <h6>
        Filtrar por:
      </h6>
      <div
        class="row"
      >
        <div
          class="col s12 m4"
        >
          <select
            class="browser-default"
            name="language"
          >
            <option
              disabled=""
              selected=""
              value=""
            >
              Idioma
            </option>
            <option
              value="Pashto"
            >
              Pashto
            </option>
            <option
              value="Uzbek"
            >
              Uzbek
            </option>
            <option
              value="Turkmen"
            >
              Turkmen
            </option>
            <option
              value="Swedish"
            >
              Swedish
            </option>
            <option
              value="Albanian"
            >
              Albanian
            </option>
            <option
              value="Arabic"
            >
              Arabic
            </option>
            <option
              value="English"
            >
              English
            </option>
            <option
              value="Samoan"
            >
              Samoan
            </option>
            <option
              value="Catalan"
            >
              Catalan
            </option>
            <option
              value="Portuguese"
            >
              Portuguese
            </option>
          </select>
        </div>
        <div
          class="col s12 m4"
        >
          <select
            class="browser-default"
            name="currency"
          >
            <option
              disabled=""
              selected=""
              value=""
            >
              Moneda
            </option>
            <option
              value="Afghan afghani"
            >
              Afghan afghani
            </option>
            <option
              value="Euro"
            >
              Euro
            </option>
            <option
              value="Albanian lek"
            >
              Albanian lek
            </option>
            <option
              value="Algerian dinar"
            >
              Algerian dinar
            </option>
            <option
              value="United State Dollar"
            >
              United State Dollar
            </option>
            <option
              value="Angolan kwanza"
            >
              Angolan kwanza
            </option>
            <option
              value="East Caribbean dollar"
            >
              East Caribbean dollar
            </option>
            <option
              value="Australian dollar"
            >
              Australian dollar
            </option>
            <option
              value="British pound"
            >
              British pound
            </option>
          </select>
        </div>
        <div
          class="col s12 m4"
        >
          <select
            class="browser-default"
            name="region"
          >
            <option
              disabled=""
              selected=""
              value=""
            >
              Región
            </option>
            <option
              value="Asia"
            >
              Asia
            </option>
            <option
              value="Europe"
            >
              Europe
            </option>
            <option
              value="Africa"
            >
              Africa
            </option>
            <option
              value="Oceania"
            >
              Oceania
            </option>
            <option
              value="Americas"
            >
              Americas
            </option>
            <option
              value="Polar"
            >
              Polar
            </option>
          </select>
        </div>
      </div>
      <button
        class="btn waves-effect waves-light red"
        name="action"
        type="submit"
      >
        Limpiar
      </button>
      <div
        class="row"
      >
        <div
          class="preloader-wrapper big active"
        >
          <div
            class="spinner-layer spinner-blue-only"
          >
            <div
              class="circle-clipper left"
            >
              <div
                class="circle"
              />
            </div>
          </div>
        </div>
        <label>
          No se encontraron resultados.
        </label>
      </div>
    </div>
  </div>
</div>
`;
